services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/QueueAttributesManager.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.sqs.internal.batchmanager;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.services.sqs.SqsAsyncClient;
+import software.amazon.awssdk.services.sqs.model.GetQueueAttributesRequest;
+import software.amazon.awssdk.services.sqs.model.QueueAttributeName;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
+import software.amazon.awssdk.utils.Validate;
+
+
+@SdkInternalApi
+public class QueueAttributesManager {
+
+    private final SqsAsyncClient sqsClient;
+    private final String queueUrl;
+    private final AtomicReference<CompletableFuture<Duration>> defaultWaitTimeSecondsFuture = new AtomicReference<>();
+    private final AtomicReference<CompletableFuture<Duration>> visibilityTimeoutSecondsFuture = new AtomicReference<>();
+    private final Duration minReceiveWaitTime;
+
+    public QueueAttributesManager(SqsAsyncClient sqsClient, String queueUrl, Duration minReceiveWaitTime) {
+        this.sqsClient = sqsClient;
+        this.queueUrl = queueUrl;
+        this.minReceiveWaitTime = minReceiveWaitTime;
+    }
+
+    /**
+     * Retrieves the received message timeout based on the provided request and queue attributes.
+     *
+     * @param rq The receive message request
+     * @return CompletableFuture with the calculated receive message timeout in milliseconds
+     */
+    public CompletableFuture<Duration> getReceiveMessageTimeout(ReceiveMessageRequest rq) {
```

Comment:
Should we short circuit here if `rq` has a custom wait time set? If so we don't need to know the default value right?

Reply:
yes, seems like we can just use `volatile Duration` for waitTimeSeconds and do something like the following


```
    public CompletableFuture<Duration> getReceiveMessageTimeout(ReceiveMessageRequest request) {
        if (request.waitTimeSeconds() != null) {
            return CompletableFuture.completedFuture(Duration.ofSeconds(request.waitTimeSeconds()));
        }

        if (defaultWaitTimeSecondsFuture == null) {
            synchronized (this) {
                if (defaultWaitTimeSecondsFuture == null) {
                    GetQueueAttributesRequest req = GetQueueAttributesRequest.builder()
                                                                             .queueUrl(queueUrl)
                                                                             .attributeNames(
                                                                                 QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS)
                                                                             .build();

                    return sqsClient.getQueueAttributes(req)
                                    .thenApply(response -> {
                                        String messageWaitTime =
                                            Validate.notNull(response
                                                                 .attributes()
                                                                 .get(QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS),
                                                             QueueAttributeName.RECEIVE_MESSAGE_WAIT_TIME_SECONDS +
                                                             " attribute is null in sqs.");

                                        Duration duration = Duration.ofSeconds(Integer.parseInt(messageWaitTime));
                                        defaultWaitTimeSecondsFuture = duration;
                                        return duration;
                                    });
                }
            }
        }
        
        return CompletableFuture.completedFuture(defaultWaitTimeSecondsFuture);
    }
```

Reply:
No,  we need that because we take the maximum of the value 
https://github.com/aws/aws-sdk-java/blob/9acdd7a744fef85a56ca298a955890dce6111688/aws-java-sdk-sqs/src/main/java/com/amazonaws/services/sqs/buffered/ReceiveQueueBuffer.java#L182-L197

In order to determin max we have to pull the Value as configured on actual queue

Reply:
good catch